proposals/threads/DataSegmentInitialization.md

```diff
@@ -0,0 +1,135 @@
+# Data segment initialization flag proposal
+
+This page describes a proposal for providing a flag to skip data segment
+initialization when instantiating a module.
+
+## Rationale
+
+Under the current threading proposal, to share a module between multiple
+agents, the module must be instantiated multiple times: once per agent.
+Instantiation initializes linear memory with the contents in the module's data
+segments. If the memory is shared between multiple agents, it will be
+initialized multiple times, potentially overwriting stores that occurred after
+the previous initializations.
+
+For example:
+
+```webassembly
+// The module.
+(module
+  (memory (export "memory") 1)
+
+  ;; Some value used as a counter.
+  (data (i32.const 1) "\0")
+
+  ;; Add one to the counter.
+  (func (export "addOne")
+    (i32.store8
+      (i32.const 0)
+      (i32.add
+        (i32.load8_u (i32.const 0))
+        (i32.const 1)))
+  )
+)
+```
+
+```javascript
+// main.js
+let moduleBytes = ...;
+
+WebAssembly.instantiate(moduleBytes).then(
+  ({instance}) => {
+    // Increment our counter.
+    instance.exports.addOne();
+
+    // Spawn a new Worker.
+    new Worker('worker.js');
+  });
+
+// worker.js
+let moduleBytes = ...;
```

Comment:
Mostly a nit, but if the example becomes long-lived, it'd be nice if main.js would postMessage the module to the worker which the worker would receive via onmessage.